[BUG]  ScrollId not set to null after scroll clean

```
23/11/09 07:42:21 ERROR FlintREPL: Unexpected error: OpenSearch exception [type=search_phase_execution_exception, reason=all shards failed]
org.opensearch.OpenSearchStatusException: OpenSearch exception [type=search_phase_execution_exception, reason=all shards failed]
	at org.opensearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:207) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.parseEntity(RestHighLevelClient.java:2228) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.parseResponseException(RestHighLevelClient.java:2205) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1924) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1877) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1845) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.client.RestHighLevelClient.scroll(RestHighLevelClient.java:1208) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.flint.core.storage.OpenSearchScrollReader.search(OpenSearchScrollReader.java:54) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.flint.core.storage.OpenSearchReader.hasNext(OpenSearchReader.java:41) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.apache.spark.sql.FlintREPL$.processCommands(FlintREPL.scala:280) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.apache.spark.sql.FlintREPL$.queryLoop(FlintREPL.scala:156) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.apache.spark.sql.FlintREPL$.main(FlintREPL.scala:99) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.apache.spark.sql.FlintREPL.main(FlintREPL.scala) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1006) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1095) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1104) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
	Suppressed: org.opensearch.client.ResponseException: method [POST], host [https://search-flint-loadtest-2-11-7do6qjpsombdtavf5r74zvs4dy.us-west-2.es-release.amazonaws.com], URI [/_search/scroll], status line [HTTP/1.1 404 Not Found]
{"error":{"root_cause":[{"type":"search_context_missing_exception","reason":"No search context found for id [160225]"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":-1,"index":null,"reason":{"type":"search_context_missing_exception","reason":"No search context found for id [160225]"}}],"caused_by":{"type":"search_context_missing_exception","reason":"No search context found for id [160225]"}},"status":404}
		at org.opensearch.client.RestClient.convertResponse(RestClient.java:375) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestClient.performRequest(RestClient.java:345) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestClient.performRequest(RestClient.java:320) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1911) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1877) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestHighLevelClient.performRequestAndParseEntity(RestHighLevelClient.java:1845) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.client.RestHighLevelClient.scroll(RestHighLevelClient.java:1208) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.flint.core.storage.OpenSearchScrollReader.search(OpenSearchScrollReader.java:54) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.opensearch.flint.core.storage.OpenSearchReader.hasNext(OpenSearchReader.java:41) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.apache.spark.sql.FlintREPL$.processCommands(FlintREPL.scala:280) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.apache.spark.sql.FlintREPL$.queryLoop(FlintREPL.scala:156) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.apache.spark.sql.FlintREPL$.main(FlintREPL.scala:99) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at org.apache.spark.sql.FlintREPL.main(FlintREPL.scala) ~[opensearch-spark-sql-application_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
		at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
		at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
		at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
		at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
		at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:1006) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1095) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1104) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
		at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala) ~[spark-core_2.12-3.3.2-amzn-0.jar:3.3.2-amzn-0]
Caused by: org.opensearch.OpenSearchException: OpenSearch exception [type=search_context_missing_exception, reason=No search context found for id [160225]]
	at org.opensearch.OpenSearchException.innerFromXContent(OpenSearchException.java:548) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.OpenSearchException.fromXContent(OpenSearchException.java:459) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.OpenSearchException.innerFromXContent(OpenSearchException.java:489) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.OpenSearchException.failureFromXContent(OpenSearchException.java:653) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	at org.opensearch.rest.BytesRestResponse.errorFromXContent(BytesRestResponse.java:199) ~[opensearch-spark-standalone_2.12-0.1.0.jar:0.1.0-SNAPSHOT]
	... 24 more
```

--- a/docs/index.md
+++ b/docs/index.md
@@ -357,6 +357,7 @@ In the index mapping, the `_meta` and `properties`field stores meta and schema i
 - `spark.datasource.flint.write.refresh_policy`: default value is wait_for. valid values [NONE
   (false), IMMEDIATE(true), WAIT_UNTIL(wait_for)]
 - `spark.datasource.flint.read.scroll_size`: default value is 100.
+- `spark.datasource.flint.read.scroll_duration`: default value is 5 minutes. scroll context keep alive duration.
 - `spark.flint.optimizer.enabled`: default is true.
 - `spark.flint.index.hybridscan.enabled`: default is false.
 - `spark.flint.index.checkpoint.mandatory`: default is true.

--- a/flint-core/src/main/scala/org/opensearch/flint/core/FlintOptions.java
+++ b/flint-core/src/main/scala/org/opensearch/flint/core/FlintOptions.java
@@ -50,6 +50,12 @@ public class FlintOptions implements Serializable {
   public static final String SCROLL_SIZE = "read.scroll_size";
   public static final int DEFAULT_SCROLL_SIZE = 100;
 
+  public static final String SCROLL_DURATION = "read.scroll_duration";
+  /**
+   * 5 minutes;
+   */
+  public static final int DEFAULT_SCROLL_DURATION = 5;
+
   public static final String REFRESH_POLICY = "write.refresh_policy";
   /**
    * NONE("false")
@@ -74,6 +80,10 @@ public class FlintOptions implements Serializable {
 
   public int getScrollSize() {
     return Integer.parseInt(options.getOrDefault(SCROLL_SIZE, String.valueOf(DEFAULT_SCROLL_SIZE)));
+  }
+
+  public int getScrollDuration() {
+    return Integer.parseInt(options.getOrDefault(SCROLL_DURATION, String.valueOf(DEFAULT_SCROLL_DURATION)));
   }
 
   public String getRefreshPolicy() {return options.getOrDefault(REFRESH_POLICY, DEFAULT_REFRESH_POLICY);}

--- a/flint-core/src/main/scala/org/opensearch/flint/core/storage/OpenSearchReader.java
+++ b/flint-core/src/main/scala/org/opensearch/flint/core/storage/OpenSearchReader.java
@@ -14,6 +14,7 @@ import java.io.IOException;
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Optional;
 
 /**
  * Abstract OpenSearch Reader.
@@ -38,8 +39,12 @@ public abstract class OpenSearchReader implements FlintReader {
   @Override public boolean hasNext() {
     try {
       if (iterator == null || !iterator.hasNext()) {
-        SearchResponse response = search(searchRequest);
-        List<SearchHit> searchHits = Arrays.asList(response.getHits().getHits());
+        Optional<SearchResponse> response = search(searchRequest);
+        if (response.isEmpty()) {
+          iterator = null;
+          return false;
+        }
+        List<SearchHit> searchHits = Arrays.asList(response.get().getHits().getHits());
         iterator = searchHits.iterator();
       }
       return iterator.hasNext();
@@ -72,7 +77,7 @@ public abstract class OpenSearchReader implements FlintReader {
   /**
    * search.
    */
-  abstract SearchResponse search(SearchRequest request) throws IOException;
+  abstract Optional<SearchResponse> search(SearchRequest request) throws IOException;
 
   /**
    * clean.

--- a/flint-core/src/main/scala/org/opensearch/flint/core/storage/OpenSearchScrollReader.java
+++ b/flint-core/src/main/scala/org/opensearch/flint/core/storage/OpenSearchScrollReader.java
@@ -18,6 +18,7 @@ import org.opensearch.flint.core.FlintOptions;
 import org.opensearch.search.builder.SearchSourceBuilder;
 
 import java.io.IOException;
+import java.util.Optional;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
@@ -28,30 +29,37 @@ public class OpenSearchScrollReader extends OpenSearchReader {
 
   private static final Logger LOG = Logger.getLogger(OpenSearchScrollReader.class.getName());
 
-  /** Default scroll context timeout in minutes. */
-  public static final TimeValue DEFAULT_SCROLL_TIMEOUT = TimeValue.timeValueMinutes(5L);
-
   private final FlintOptions options;
+
+  private final TimeValue scrollDuration;
 
   private String scrollId = null;
 
   public OpenSearchScrollReader(RestHighLevelClient client, String indexName, SearchSourceBuilder searchSourceBuilder, FlintOptions options) {
     super(client, new SearchRequest().indices(indexName).source(searchSourceBuilder.size(options.getScrollSize())));
     this.options = options;
+    this.scrollDuration = TimeValue.timeValueMinutes(options.getScrollDuration());
   }
 
   /**
    * search.
    */
-  SearchResponse search(SearchRequest request) throws IOException {
+  Optional<SearchResponse> search(SearchRequest request) throws IOException {
     if (Strings.isNullOrEmpty(scrollId)) {
-      // add scroll timeout making the request as scroll search request.
-      request.scroll(DEFAULT_SCROLL_TIMEOUT);
+      request.scroll(scrollDuration);
       SearchResponse response = client.search(request, RequestOptions.DEFAULT);
       scrollId = response.getScrollId();
-      return response;
+      return Optional.of(response);
     } else {
-      return client.scroll(new SearchScrollRequest().scroll(DEFAULT_SCROLL_TIMEOUT).scrollId(scrollId), RequestOptions.DEFAULT);
+      try {
+        return Optional
+        .of(client.scroll(new SearchScrollRequest().scroll(scrollDuration).scrollId(scrollId),
+            RequestOptions.DEFAULT));
+      } catch (OpenSearchStatusException e) {
+        LOG.log(Level.WARNING, "scroll context not exist", e);
+        scrollId = null;
+        return Optional.empty();
+      }
     }
   }
 
@@ -69,6 +77,15 @@ public class OpenSearchScrollReader extends OpenSearchReader {
       // OpenSearch throw exception if scroll already closed. https://github.com/opensearch-project/OpenSearch/issues/11121
       LOG.log(Level.WARNING, "close scroll exception, it is a known bug https://github" +
           ".com/opensearch-project/OpenSearch/issues/11121.", e);
+    } finally {
+      scrollId = null;
     }
+  }
+
+  /**
+   * Public for testing.
+   */
+  public String getScrollId() {
+    return scrollId;
   }
 }

--- a/flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala
+++ b/flint-spark-integration/src/main/scala/org/apache/spark/sql/flint/config/FlintSparkConf.scala
@@ -106,6 +106,11 @@ object FlintSparkConf {
     .doc("scroll read size")
     .createWithDefault("100")
 
+  val SCROLL_DURATION = FlintConfig(s"spark.datasource.flint.${FlintOptions.SCROLL_DURATION}")
+    .datasourceOption()
+    .doc("scroll duration in minutes")
+    .createWithDefault(String.valueOf(FlintOptions.DEFAULT_SCROLL_DURATION))
+
   val OPTIMIZER_RULE_ENABLED = FlintConfig("spark.flint.optimizer.enabled")
     .doc("Enable Flint optimizer rule for query rewrite with Flint index")
     .createWithDefault("true")
@@ -158,6 +163,7 @@ case class FlintSparkConf(properties: JMap[String, String]) extends Serializable
         HOST_PORT,
         REFRESH_POLICY,
         SCROLL_SIZE,
+        SCROLL_DURATION,
         SCHEME,
         AUTH,
         REGION,

--- a/integ-test/src/test/scala/org/opensearch/flint/core/FlintOpenSearchClientSuite.scala
+++ b/integ-test/src/test/scala/org/opensearch/flint/core/FlintOpenSearchClientSuite.scala
@@ -17,12 +17,12 @@ import org.opensearch.client.transport.rest_client.RestClientTransport
 import org.opensearch.flint.OpenSearchSuite
 import org.opensearch.flint.core.metadata.FlintMetadata
 import org.opensearch.flint.core.metadata.log.OptimisticTransaction.NoOptimisticTransaction
-import org.opensearch.flint.core.storage.FlintOpenSearchClient
+import org.opensearch.flint.core.storage.{FlintOpenSearchClient, OpenSearchScrollReader}
 import org.scalatest.flatspec.AnyFlatSpec
 import org.scalatest.matchers.should.Matchers
 import org.scalatestplus.mockito.MockitoSugar.mock
 
-import org.apache.spark.sql.flint.config.FlintSparkConf.REFRESH_POLICY
+import org.apache.spark.sql.flint.config.FlintSparkConf.{REFRESH_POLICY, SCROLL_DURATION, SCROLL_SIZE, SCROLL_TIMEOUT}
 
 class FlintOpenSearchClientSuite extends AnyFlatSpec with OpenSearchSuite with Matchers {
 
@@ -176,6 +176,30 @@ class FlintOpenSearchClientSuite extends AnyFlatSpec with OpenSearchSuite with M
       reader.hasNext shouldBe false
       reader.close()
 
+      reader.asInstanceOf[OpenSearchScrollReader].getScrollId shouldBe null
+      scrollShouldClosed()
+    }
+  }
+
+  it should "no item return after scroll timeout" in {
+    val indexName = "t0001"
+    withIndexName(indexName) {
+      multipleDocIndex(indexName, 2)
+
+      val options =
+        openSearchOptions + (s"${SCROLL_DURATION.optionKey}" -> "1", s"${SCROLL_SIZE.optionKey}" -> "1")
+      val flintClient = new FlintOpenSearchClient(new FlintOptions(options.asJava))
+      val match_all = null
+      val reader = flintClient.createReader(indexName, match_all)
+
+      reader.hasNext shouldBe true
+      reader.next
+      // scroll context expired after 1 minutes
+      Thread.sleep(60 * 1000 * 2)
+      reader.hasNext shouldBe false
+      reader.close()
+
+      reader.asInstanceOf[OpenSearchScrollReader].getScrollId shouldBe null
       scrollShouldClosed()
     }
   }

--- a/integ-test/src/test/scala/org/opensearch/flint/core/FlintOpenSearchClientSuite.scala
+++ b/integ-test/src/test/scala/org/opensearch/flint/core/FlintOpenSearchClientSuite.scala
@@ -22,7 +22,7 @@ import org.scalatest.flatspec.AnyFlatSpec
 import org.scalatest.matchers.should.Matchers
 import org.scalatestplus.mockito.MockitoSugar.mock
 
-import org.apache.spark.sql.flint.config.FlintSparkConf.{REFRESH_POLICY, SCROLL_DURATION, SCROLL_SIZE, SCROLL_TIMEOUT}
+import org.apache.spark.sql.flint.config.FlintSparkConf.{REFRESH_POLICY, SCROLL_DURATION, SCROLL_SIZE}
 
 class FlintOpenSearchClientSuite extends AnyFlatSpec with OpenSearchSuite with Matchers {
 

